sycl/doc/extensions/proposed/sycl_ext_oneapi_joint_for.asciidoc

```diff
@@ -0,0 +1,210 @@
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+A very common idiom in SPMD programming sees developers distribute a range over
+the members of a group, by writing a loop that looks like this:
+
+[source,c++]
+----
+for (int i = it.get_local_id(0); i < N; i += it.get_local_range(0)) { ... }
+----
+
+The combination of setting the initial value based on the work-item index and
+setting the stride based on the number of work-items in the group ensures that
+any data accesses within the loop are appropriately interleaved to maximize
+performance.
+
+Although common, this idiom has several drawbacks: it is difficult to teach to
+developers unfamiliar with SPMD programming; it is easy to use incorrectly
+(e.g., by forgetting to set a different initial value on each work-item); and
+it is difficult to use generically (i.e., for different groups of work-items).
+
+This extension proposes to encapsulate this pattern in group algorithms, to
+simplify generic usage across different group types.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_JOINT_FOR` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Group algorithms
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template <typename Group, typename InputIterator, typename Function>
+Function joint_for_each(Group g, InputIterator first, InputIterator last, Function f);
+
+} // namespace sycl::ext::oneapi
+----
+
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>>` is
+true and `InputIterator` is a random access iterator.
+
+_Preconditions_: `first` and `last` must be the same for all work-items in
+group `g`, and `f` must be an immutable callable with the same type and state
+for all work-items in group `g`.
+
+_Effects_: Blocks until all work-items in group `g` have
+reached this synchronization point, then applies `f` to the result of
+dereferencing every iterator in the range [`first`, `last`).
+The range is distributed across the work-items in group `g`.
+
+_Synchronization_: The call to this function in each work-item
+happens before the algorithm begins execution.
+The completion of the algorithm happens before any work-item
+blocking on the same synchronization point is unblocked.
+
+_Remarks_: If `f` returns a result, the result is ignored.
```

Comment:
Why not add a constraint that the return value must be `void`?

For that matter, are any parameters passed to the callable?  If not, the should also be a constraint that the callable can be invoked with no parameters.

Reply:
I didn't add constraints like these because `std::for_each` doesn't have them: https://eel.is/c++draft/alg.foreach.

Do you still think we should add these extra constraints?

Reply:
OK, makes sense to match C++.

I think it would be good to add an example, though, of both `joint_for_each` and `joint_for`.  This could help illustrate the parameter of the callable.  For example, it might not be obvious that the parameter can be a reference which the callable can use to modify the element.

Reply:
Good idea.  Added examples in 1c3c61f69244c53a8d7b9d7b3261646b368ea755.